homeassistant/components/bsblan/config_flow.py

```diff
@@ -0,0 +1,137 @@
+from homeassistant.const import CONF_HOST, CONF_NAME, CONF_PORT
+from homeassistant.helpers import ConfigType
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+
+from .const import CONF_DEVICE_IDENT, DOMAIN  # pylint:disable=unused-import
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class BSBLanFlowHandler(ConfigFlow, domain=DOMAIN):
+    """Handle a BSBLan config flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = CONN_CLASS_LOCAL_POLL
+
+    async def async_step_user(
+        self, user_input: Optional[ConfigType] = None
+    ) -> Dict[str, Any]:
+        """Handle a flow initiated by the user."""
+        if user_input is None:
+            return self._show_setup_form()
+
+        try:
+            info = await self._get_bsblan_info(
+                user_input[CONF_HOST], user_input[CONF_PORT]
+            )
+        except BSBLanError:
+            return self._show_setup_form({"base": "connection_error"})
+
+        # Check if already configured
+        await self.async_set_unique_id(info.device_identification)
+        self._abort_if_unique_id_configured()
+
+        return self.async_create_entry(
+            title=info.device_identification,
+            data={
+                CONF_HOST: user_input[CONF_HOST],
+                CONF_PORT: user_input[CONF_PORT],
+                CONF_DEVICE_IDENT: info.device_identification,
+            },
+        )
+
+    async def async_step_zeroconf(
+        self, user_input: Optional[ConfigType] = None
+    ) -> Dict[str, Any]:
+        """Handle zeroconf discovery."""
+        if user_input is None:
+            return self.async_abort(reason="connection_error")
+
+        # Hostname is format: my-ke.local.
+        host = user_input["hostname"].rstrip(".")
+        try:
+            info = await self._get_bsblan_info(host, user_input[CONF_PORT])
+        except BSBLanError:
+            return self.async_abort(reason="connection_error")
+
+        # Check if already configured
+        await self.async_set_unique_id(info.device_identification)
+        self._abort_if_unique_id_configured()
+
+        # pylint: disable=no-member # https://github.com/PyCQA/pylint/issues/3167
+        self.context.update(
+            {
+                CONF_HOST: host,
+                CONF_PORT: user_input[CONF_PORT],
+                CONF_DEVICE_IDENT: info.device_identification,
+                "title_placeholders": {"device_ident": info.device_identification},
+            }
+        )
+
+        # Prepare configuration flow
+        return self._show_confirm_dialog()
+
+    # pylint: disable=no-member # https://github.com/PyCQA/pylint/issues/3167
+    async def async_step_zeroconf_confirm(
+        self, user_input: ConfigType = None
+    ) -> Dict[str, Any]:
+        """Handle a flow initiated by zeroconf."""
+        if user_input is None:
+            return self._show_confirm_dialog()
+
+        try:
+            info = await self._get_bsblan_info(
```

Comment:
We already do this in the previous step. Do we need to do it again? Can we store the info as an instance attribute instead?